_posts/2019-04-02-running-IMA-on-keylime.markdown

```diff
@@ -0,0 +1,441 @@
+author: "Luke Hinds"
+date:   2019-04-02
+---
+
+A feature of Keylime is runtime integrity monitoring, using the integrity management architecture (IMA).
+
+## Intro to IMA
+
+IMA was introduced to the Linux Kernel on the 20th May 2005 for the Kernel release 2.6.30 as part of the overall
+Linux Integrity Subsystem. The design was based upon the work of the open standard for trusted compute by the Trusted Computing Group (TCG).
+
+IMA (also known as the kernel integrity subsystem) provides a means of detecting whether files have been accidentally or maliciously altered or executed both remotely or locally.
+
+The following modules provide the integrity functions:
+
+* Collect – measure a file before it is accessed.
+* Store – add the measurement to a kernel resident list and, if a
+hardware Trusted Platform Module (TPM) is present, extend the IMA
+PCR
+* Attest – if present, use the TPM to sign the IMA PCR value, to
+allow a remote validation of the measurement list.
+* Appraise – enforce local validation of a measurement against a
+“good” value stored in an extended attribute of the file.
+* Protect – protect a file's security extended attributes
+(including appraisal hash) against off-line attack.
+* Audit – audit the file hashes.
+
+Keylime makes use of the attest function "IMA-measurement"
+
+## IMA-measurement
+
+IMA maintains a runtime measurement list and, if a hardware Trusted Platform Module (TPM) is present, the list is anchored within the Trusted Platform Modules (TPM) Platform Configuration Registers (`PCR`). The benefit of anchoring the aggregate integrity value in the TPM is that the measurement list cannot be compromised by any software attack, without being detectable. Hence, on a trusted boot system, IMA-measurement can be used to attest to the system's runtime integrity remotely (or locally).
+
+A majority of Linux distributions now have IMA measurement enabled by default. For this article we will use Fedora 29, but the feature is available in many earlier versions of the RHEL and Debian family, alongside Gentoo, Arch Linux and OpenSUSE.
+
+Quite often, all that is required is an `ima-policy`. This is then placed into `/etc/ima/ima-policy`. In keylime, for demo purposes we provide the following `ima-policy`
+
+```
+# PROC_SUPER_MAGIC
+dont_measure fsmagic=0x9fa0
+# SYSFS_MAGIC
+dont_measure fsmagic=0x62656572
+# DEBUGFS_MAGIC
+dont_measure fsmagic=0x64626720
+# TMPFS_MAGIC
+dont_measure fsmagic=0x01021994
+# RAMFS_MAGIC
+dont_measure fsmagic=0x858458f6
+# SECURITYFS_MAGIC
+dont_measure fsmagic=0x73636673
+# MEASUREMENTS
+measure func=BPRM_CHECK
+measure func=FILE_MMAP mask=MAY_EXEC
+measure func=MODULE_CHECK uid=0
+```
+
+This default policy measures all executables in `bprm_check`, all files mmapped executable in `file_mmap`, and all files
+open for read by `root` in `do_filp_open`.
+
+It is of course possible to create your own IMA policy, along with LSM specific definitions for `SELinux` or `Smack`.
```

Comment:
Not sure if we want to include this explicitly, but one example of using SELinux here is to have IMA watch all files in /etc/: 
`measure func=FILE_CHECK mask=MAY_READ obj_type=etc_t`

Reply:
Nice! I added that